### DISCUSSION TITLE
Running commands asynchronously with asyncio.subprocess breaks Ghostty

### DISCUSSION THREAD

**Discussion by hankertrix** [ghostty-org/ghostty#3472] [Issue Triage] [answered]
After running multiple long-running commands asynchronously with Python's asyncio.subprocess, Ghostty no longer displays anything that is typed to the terminal.

Below is the file with the script, it requires `python`, `ffmpeg` and [`agg`](https://github.com/asciinema/agg) to be present on your PATH.

[reproduction.zip](https://github.com/user-attachments/files/18261963/reproduction.zip)

Changing the script to output to the null output still results in the same issue, while changing the script to run synchronously does not cause any issues with Ghostty.

**Comment by 00-kat** [comment]
Unsolicited code review: [pathlib.PurePath.with_suffix()] can be used instead of whatever you're doing with `str(something) + ".ext"`.

Notes for people trying to reproduce this:
  - There are [AGG binaries over at their GitHub](https://github.com/asciinema/agg/releases), if you understandably don't want to waste your resources compiling rust.
  - If you grab a binary, make sure you `chmod` it, otherwise you will get strange errors.

The script finished for me successfully, and I could not reproduce this.  My guess, however, is either:
  - There *was* a bug, but it was fixed between you reporting this and me finding this.
  - AGG's progress bar has to manipulate the terminal, and a lot of them running simultaneously is resulting in them stepping on each others' feet.  I noticed that the progress bar looked odd and there were multiple drawn at times and other similar issues.  While this would be my guess, running `python generate_videos.py >/dev/null 2>/dev/null` *shouldn't* have resulted in any issues, unless it was also modifying the terminal I/O structure[^1] ([termios]) for whatever reason.  You can use the terminal inspector (same default keybinding as your browser's inspector) to see if it has changed anything which could cause issues (though the inspector notably and unfortunately does not include the [termios] state).

[pathlib.PurePath.with_suffix()]: https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.with_suffix
[termios]: https://www.man7.org/linux/man-pages/man3/termios.3.html

[^1]: Not sure if it's "terminal I/O structure" or "terminal I/O state", I couldn't find an expansion in the documentation.

**Reply by hankertrix** [reply]
> Unsolicited code review: [pathlib.PurePath.with_suffix()](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.with_suffix) can be used instead of whatever you're doing with `str(something) + ".ext"`

Thanks, I'll keep that in mind for future scripts.

> The script finished for me successfully, and I could not reproduce this.

Hmm it is still an issue for me. However, it also happens in pretty much every other terminal that I use, like `kitty` and `wezterm`.

Below is a screenshot of the terminal IO when I run the script, piping the output to `/dev/null`:

![image](https://github.com/user-attachments/assets/50cd3b87-e1c7-47a3-9512-5e42741afc96)

**Reply by rhodes-b** [reply]
> Hmm it is still an issue for me. However, it also happens in pretty much every other terminal that I use, like kitty and wezterm.

Then this is not a ghostty issue? Rather something with your environment

**Reply by hankertrix** [reply]
I’m not sure what else in my environment could be causing this issue though, since it is related to the display of the terminal.

The shell still works properly, it’s just the text isn’t being displayed.

**Reply by rhodes-b** [reply]
If it reproduces in all the other terminals you tested its not ghosttys fault though which is what discussions are for

Like stated kat could not reproduce this which means its heavily infered as a environment issue

We cant know what about your environment causes this. Its not our setup

**Reply by hankertrix** [reply]
@00-kat, may I get the output of `stty --all` before and after you run the script, and preferably the versions of `python` and `ffmpeg` you have?

This seems to be an issue with either `agg`, `ffmpeg` or `python` changing the terminal settings.

Source:
https://askubuntu.com/questions/171449/shell-does-not-show-typed-in-commands-reset-works-but-what-happened

My `python` version is `3.13.1` and my `ffmpeg` version is `n7.1`.

Here's the output of `stty --all` from before I run the command.

```
speed 38400 baud; rows 50; columns 167; line = 0;
intr = ^C; quit = ^\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>; eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R; werase = ^W;
lnext = ^V; discard = ^O; min = 1; time = 0;
-parenb -parodd -cmspar cs8 -hupcl -cstopb cread -clocal -crtscts
-ignbrk brkint ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl ixon -ixoff -iuclc -ixany imaxbel iutf8
opost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig icanon -iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt echoctl echoke -flusho -extproc
```

Here's the output of `stty --all` after running the script:

```
speed 38400 baud; rows 50; columns 167; line = 0;
intr = ^C; quit = ^\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>; eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R; werase = ^W;
lnext = ^V; discard = ^O; min = 1; time = 0;
-parenb -parodd -cmspar cs8 -hupcl -cstopb cread -clocal -crtscts
-ignbrk -brkint ignpar -parmrk -inpck -istrip -inlcr -igncr -icrnl -ixon -ixoff -iuclc -ixany imaxbel iutf8
opost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig -icanon -iexten -echo echoe echok -echonl -noflsh -xcase -tostop -echoprt echoctl echoke -flusho -extproc
```

Diffing the two:

```diff
5c5
< -ignbrk brkint ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl ixon -ixoff -iuclc -ixany imaxbel iutf8
---
> -ignbrk -brkint ignpar -parmrk -inpck -istrip -inlcr -igncr -icrnl -ixon -ixoff -iuclc -ixany imaxbel iutf8
7c7
< isig icanon -iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt echoctl echoke -flusho -extproc
---
> isig -icanon -iexten -echo echoe echok -echonl -noflsh -xcase -tostop -echoprt echoctl echoke -flusho -extproc
```

**Reply by 00-kat** [reply, marked as answer]
> This seems to be an issue with either `agg`, `ffmpeg` or `python` changing the terminal settings.

That was my guess too, yeah (in my second bullet point).  I'm thinking `agg` is the one touching it for the progress bar (they also overwrite each other continually), and when you run them in parallel they're stepping on each others' toes, so it doesn't get reset properly.  I doubt Python touches the terminal (except for tracebacks in 3.13 onward) and I also doubt ffmpeg does much more than changing the colour of text.

Python 3.13.1 and ffmpeg n7.1 from the Arch Linux repos, AGG 1.5.0 from [this specific binary on their GitHub releases page](https://github.com/asciinema/agg/releases/download/v1.5.0/agg-x86_64-unknown-linux-gnu).  Running with `PATH="$PATH:." python generate_videos.py` since I'm lazy.

Before:
```console
$ stty --all
speed 38400 baud; rows 36; columns 145; line = 0;
intr = ^C; quit = ^\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>; eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z;
rprnt = ^R; werase = ^W; lnext = ^V; discard = ^O; min = 1; time = 0;
-parenb -parodd -cmspar cs8 -hupcl -cstopb cread -clocal -crtscts
-ignbrk -brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl -ixon -ixoff -iuclc -ixany -imaxbel iutf8
opost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig icanon iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt echoctl echoke -flusho -extproc
```

After: identical.

Perhaps they finish at different times for you, possibly because you have a better/worse processor or something, which results in you seeing the issue as it doesn't get reset properly, while it does for me.

**EDIT**: I forgot but Fish resets a lot of the terminal state after commands finish, which may be why it's the same for me, I'll edit the results of trying this in Dash (which doesn't touch the terminal *whatsoever*) and see if I can reproduce this.

Edit 2: Well I can definitely reproduce this with Dash, but I can't type anymore (it's not that it doesn't show up, Dash doesn't read it either) so I can't get you the diff...

**Reply by 00-kat** [reply]
Oh and this would be the issue:
> `-echo`

That flag basically says "don't display anything please!".

You can right click anywhere and click on "reset" to fix it when it breaks I think (edit: never mind, it doesn't work), and you can also try blindly run `stty sane` once it exits.

**Reply by 00-kat** [reply]
Forgot to comment on this one:

> Below is a screenshot of the terminal IO when I run the script, piping the output to /dev/null:

30 through 39 are from shell integration, 29 is from your shell enabling bracketed paste, and the rest look like they're use for the progress bar to move the cursor back, though that's just a guess.  Did you pipe both stdout and stderr to `/dev/null` by the way?  Since CSI and OSC don't work when you can't get the terminal to see them.

**Reply by hankertrix** [reply]
> I'm thinking `agg` is the one touching it for the progress bar (they also overwrite each other continually), and when you run them in parallel they're stepping on each others' toes, so it doesn't get reset properly.

Yeahhh this seems to be the most likely case. Looks like I should file a bug report there.

> I forgot but Fish resets a lot of the terminal state after commands finish, which may be why it's the same for me, I'll edit the results of trying this in Dash (which doesn't touch the terminal _whatsoever_) and see if I can reproduce this.

> Well I can definitely reproduce this with Dash, but I can't type anymore (it's not that it doesn't show up, Dash doesn't read it either) so I can't get you the diff...

I didn’t even think the shell would make a difference. I’m using `bash`, which seems like it does do something to the terminal after the script has completed, as I can still type in the terminal.

> That flag basically says "don't display anything please!".
> 
> You can right click anywhere and click on "reset" to fix it when it breaks I think (edit: never mind, it doesn't work), and you can also try blindly run `stty sane` once it exits.

Thanks for the tip! I used to just run `reset` to fix it.

> 30 through 39 are from shell integration, 29 is from your shell enabling bracketed paste, and the rest look like they're use for the progress bar to move the cursor back, though that's just a guess.

I see, thanks for the info!

> Did you pipe both stdout and stderr to `/dev/null` by the way?

Yeap I did, I just copied the command you had in the original post.
